WordPress/src/main/java/org/wordpress/android/GCMMessageService.java

```diff
@@ -335,8 +366,55 @@ private void showNotificationForBuilder(NotificationCompat.Builder builder, Cont
     }
 
     // Clear all notifications
-    private void handleBadgeResetPN() {
-        removeAllNotifications(this);
+    private void handleBadgeResetPN(Bundle data) {
+        if (data != null && data.containsKey(PUSH_ARG_NOTE_ID)) {
+            removeNotificationWithNoteIdFromSystemBar(this, data);
+            //now that we cleared the specific notif, we can check and make any visual updates
+            if (sActiveNotificationsMap.size() > 0) {
+                Bitmap largeIconBitmap = null;
+                // here notify the existing group notification by eliminating the line that is now gone
+                String title = StringEscapeUtils.unescapeHtml(data.getString(PUSH_ARG_TITLE));
+                if (title == null) {
+                    title = getString(R.string.app_name);
+                }
+                String message = StringEscapeUtils.unescapeHtml(data.getString(PUSH_ARG_MSG));
+
+                if (sActiveNotificationsMap.size() == 1) {
+                    //only one notification remains, so get the proper message for it and re-instate in the system dashboard
+                    Bundle remainingNote = sActiveNotificationsMap.values().iterator().next();
+                    if (remainingNote != null) {
+                        String remainingNoteTitle = StringEscapeUtils.unescapeHtml(remainingNote.getString(PUSH_ARG_TITLE));
+                        if (!TextUtils.isEmpty(remainingNoteTitle)) {
+                            title = remainingNoteTitle;
+                        }
+                        String remainingNoteMessage = StringEscapeUtils.unescapeHtml(remainingNote.getString(PUSH_ARG_MSG));
+                        if (!TextUtils.isEmpty(remainingNoteMessage)) {
+                            message = remainingNoteMessage;
+                        }
+                        largeIconBitmap = getLargeIconBitmap(remainingNote.getString("icon"),
+                                shouldCircularizeNoteIcon(remainingNote.getString(PUSH_ARG_TYPE)));
+                    }
+                }
+
+                NotificationCompat.Builder builder = getNotificationBuilder(title, message);
+
+                if (largeIconBitmap == null) {
+                    largeIconBitmap = getLargeIconBitmap(data.getString("icon"), shouldCircularizeNoteIcon(PUSH_TYPE_BADGE_RESET));
+                }
+
+                if (largeIconBitmap != null) {
+                    builder.setLargeIcon(largeIconBitmap);
+                }
+
+                showGroupNotificationForBuilder(builder, message);
+            }
+        } else {
+            //TODO: check if this is necessary:
```

Comment:
Special note on this: I'm intentionally commenting out the `removeAllNotifiations(this)` line as, whenever we land this functionality to prod, removing _all_  notifs at once not only won't be necessary, because the `clear all` signal is triggered when you open the Notifications tab on the web client, which will _always happen_  before opening any single notification, therefore having the `clear all` signal enabled and sent in the first place would render the `granular clear` signal unused as there won't ever be a chance when the granular PNs will be sent (unless you keep your notifications tab open on the web client all the time, which seems unlikely)